core/services/relay/evm/read/contract.go

```diff
@@ -0,0 +1,241 @@
+package read
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/google/uuid"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+// contractBinding stores read bindings and manages the common contract event filter.
+type contractBinding struct {
+	// filterRegistrar is used to manage polling filter registration for the common contract filter.
+	// The common contract filter should be used by events that share filtering args.
+	registrar *syncedFilter
+
+	// registered is used to determine if Register was called during Chain Reader service Start.
+	// This is done to avoid calling Register while the service is not running because log poller is most likely also not running.
+	registerCalled bool
+
+	// internal properties
+	name    string
+	readers map[string]Reader       // key is read name method, event or event keys used for queryKey.
+	bound   map[common.Address]bool // bound determines if address is set to the contract binding.
+	mu      sync.RWMutex
+}
+
+func newContractBinding(name string) *contractBinding {
+	return &contractBinding{
+		name:      name,
+		readers:   make(map[string]Reader),
+		bound:     make(map[common.Address]bool),
+		registrar: newSyncedFilter(),
+	}
+}
+
+// GetReaderNamed returns a reader for the provided contract name. This method is thread safe.
+func (cb *contractBinding) GetReaderNamed(name string) (Reader, bool) {
+	cb.mu.RLock()
+	defer cb.mu.RUnlock()
+
+	binding, exists := cb.readers[name]
+
+	return binding, exists
+}
+
+// AddReaderNamed adds a new reader to the contract bindings for the provided contract name. This
+// method is thread safe.
+func (cb *contractBinding) AddReaderNamed(name string, rdr Reader) {
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+
+	cb.readers[name] = rdr
+}
+
+// Bind binds contract addresses to contract binding and registers the common contract polling filter.
+func (cb *contractBinding) Bind(ctx context.Context, registrar Registrar, bindings ...common.Address) error {
+	if cb.isBound() {
+		if err := cb.Unregister(ctx, registrar); err != nil {
+			return err
+		}
+	}
+
+	for _, binding := range bindings {
+		if cb.bindingExists(binding) {
+			continue
+		}
+
+		cb.registrar.SetName(logpoller.FilterName(cb.name + "." + uuid.NewString()))
+		cb.registrar.AddAddress(binding)
+		cb.addBinding(binding)
+	}
+
+	// registerCalled could be true without the filters being set yet.
```

Comment:
This is explained above the `registerCalled` itself
```suggestion
	// registerCalled during ChainReader start
```